Hacklang property typehint support

mihai-stancu opened the pull request:
Hi,

I have implemented:
- type-hint support for Hack Properties, Getters and Setters;
- type-hint support for PHP7 Getters and Setters;
- refactored PHP5 type-hint support for Setters so that PHP5, PHP7 and Hack all use the same Extractor but with different type-info parsers;

Tested using `phpunit` -- sorry but I don't have experience with `phpspec` yet.

N.B.: PhpDocExtractor needs to be disabled for *.hh files because it expects PHP syntax and things break so I touched that extractor too, to limit its reach on parsing files.

Commit:
Hack lang property typehint support & PHP7 getter typehint support

- Implement PropertyExtractor
- Implement GetterExtractor
- Reimplement SetterExtractor
- Porting PHPSpec tests to PHPUnit tests
- Adding PHPUnit to composer.json
- Adding PHPUnit to .travis.yml
- Adding PHPUnit bootstrap with autoload

mihai-stancu commented:
I should add a GetterExtractor while I'm at it, to sniff return type hints. Since PHP 7 has return type hints on functions this should work nicely for PHP 7 as well as Hack.

mihai-stancu commented:
Hi,

I have implemented:
* type-hint support for Hack Properties, Getters and Setters;
* type-hint support for PHP7 Getters and Setters;
* refactored PHP5 type-hint support for Setters so that PHP5, PHP7 and Hack all use the same Extractor but with different type-info parsers;

Tested using `phpunit` -- sorry but I don't have experience with `phpspec` yet.

N.B.: PhpDocExtractor needs to be disabled for *.hh files because it expects PHP syntax and things break so I touched that extractor too, to limit its reach on parsing files.

mihai-stancu commented:
@dunglas bump.

dunglas commented on src/PropertyInfo/Extractors/GetterExtractor.php:
```diff
@@ -0,0 +1,32 @@
+<?php
+
+/*
+ * (c) Mihai Stancu <stancu.t.mihai@gmail.com>
```
The copyright header must be the one of the project:

```
/*
 * (c) Kévin Dunglas <dunglas@gmail.com>
 *
 * This source file is subject to the MIT license that is bundled
 * with this source code in the file LICENSE.
 */
```

(same for all files)

dunglas commented on src/PropertyInfo/Extractors/GetterExtractor.php:
```diff
@@ -0,0 +1,32 @@
+use PropertyInfo\Type;
+
+/**
+ * Setter Extractor.
```
Should be `Getter Extractor`.

dunglas commented on src/PropertyInfo/Extractors/PhpDocExtractor.php:
```diff
@@ -102,7 +102,8 @@ public function extractTypes(\ReflectionProperty $reflectionProperty)
     {
-        if (!($fileName = $reflectionClass->getFileName())) {
+        if (!($fileName = $reflectionClass->getFileName())
+        or !in_array(pathinfo($fileName, PATHINFO_EXTENSION), ['php', 'phtml'])) {
```
Replace `or` by `||` (not the same precedence, can lead to bugs).

It is better to test explicitly on hack type IMO because php files can have a lot of other extensions (`.php3`, `.php4`, `.php5`...).

mihai-stancu replied:
I have a personal preference for `and` and `or` specifically because they have lower precedence. It leads to bugs when you're not aware of it but for conformity's sake I'll flip them the other way round.

`*.hh` files are one example but PhpDoc fails to parse `*.php` files which have PHP7 syntax inside -- what about those? I could name those files `*.php7` for now and exclude them as such...

dunglas commented on src/PropertyInfo/Tests/Data/HackData.hh:
```diff
@@ -0,0 +1,433 @@
+ */
+class HackData
+{
+    protected bool $bool;
```
Can you make properties `private` instead of `protected`.

dunglas commented on src/PropertyInfo/Tests/Data/HackData.hh:
```diff
@@ -0,0 +1,433 @@
+namespace PropertyInfo\Tests\Data;
+
+/**
+ * Property Extractor.
```
The PHPDoc must be updated to explain what this class do.

mihai-stancu replied:
This class doesn't do anything it's just a dummy class used for the tests -- like `SetterDummy`. But I have updated the doc comment nonetheless.

dunglas commented on src/PropertyInfo/Tests/Data/PHP7Data.php:
```diff
@@ -0,0 +1,78 @@
+namespace PropertyInfo\Tests\Data;
+
+/**
+ * Property Extractor.
```
same here

dunglas commented on src/PropertyInfo/Tests/Data/PHP7Data.php:
```diff
@@ -0,0 +1,78 @@
+ *
+ * @author Mihai Stancu <stancu.t.mihai@gmail.com>
+ */
+class PHP7Data extends PHP5Data
```
Must be Php7Data according to our CS.

dunglas commented on src/PropertyInfo/Tests/Data/PHP7Data.php:
```diff
@@ -0,0 +1,78 @@
+ */
+class PHP7Data extends PHP5Data
+{
+    protected $bool;
```
`private` by default.

dunglas commented on src/PropertyInfo/Tests/ExtractorTest.php:
```diff
@@ -0,0 +1,71 @@
+ *
+ * @author Mihai Stancu <stancu.t.mihai@gmail.com>
+ */
+class ExtractorTest extends \PHPUnit_Framework_TestCase
```
We use PHPSpec, not PHPUnit. Can you change to it please?

dunglas commented on src/PropertyInfo/Tests/PHP5DataProvider.php:
```diff
@@ -0,0 +1,39 @@
+ *
+ * @author Mihai Stancu <stancu.t.mihai@gmail.com>
+ */
+class PHP5DataProvider
```
Must be `Php5DataProvider`

dunglas commented on src/PropertyInfo/TypeInfoParser.php:
```diff
@@ -0,0 +1,46 @@
+ *
+ * @author Mihai Stancu <stancu.t.mihai@gmail.com>
+ */
+interface TypeInfoParser
```
`TypeInfoParserInterface`

dunglas commented on src/PropertyInfo/TypeInfoParsers/HHVMTypeInfoParser.php:
```diff
@@ -0,0 +1,127 @@
+ *
+ * @author Mihai Stancu <stancu.t.mihai@gmail.com>
+ */
+class HHVMTypeInfoParser extends NativeTypeInfoParser
```
`HhvmTypeInfoParser`

dunglas commented on src/PropertyInfo/TypeInfoParsers/HHVMTypeInfoParser.php:
```diff
@@ -0,0 +1,127 @@
+    {
+        $type = new Type();
+
+        if ($info[0] === '?') {
```
yoda style (`'?' === $info[0]`)

dunglas commented on src/PropertyInfo/TypeInfoParsers/HHVMTypeInfoParser.php:
```diff
@@ -0,0 +1,127 @@
+
+        if (array_key_exists($info, static::$types)) {
+            $type->setType(static::$types[$info]);
+            $type->setCollection($info === 'array');
```
yoda style

dunglas commented on src/PropertyInfo/TypeInfoParsers/HHVMTypeInfoParser.php:
```diff
@@ -0,0 +1,127 @@
+            $info = substr($info, 1);
+        }
+
+        if (array_key_exists($info, static::$types)) {
```
Prefer `isset` in this case (performance)

dunglas commented on src/PropertyInfo/TypeInfoParsers/HHVMTypeInfoParser.php:
```diff
@@ -0,0 +1,127 @@
+            return [$type];
+        }
+
+        if (($pos = strpos($info, '<')) !== false) {
```
yoda style

dunglas commented on src/PropertyInfo/TypeInfoParsers/HHVMTypeInfoParser.php:
```diff
@@ -0,0 +1,127 @@
+            $contents = substr($info, $pos+1, -1);
+            $contents = explode(', ', $contents);
+
+            if ($container === 'array') {
```
yoda style

dunglas commented on src/PropertyInfo/TypeInfoParsers/NativeTypeInfoParser.php:
```diff
@@ -0,0 +1,73 @@
+use PropertyInfo\TypeInfoParser;
+
+/**
+ * Property Extractor.
```
phpdoc

dunglas commented on src/PropertyInfo/TypeInfoParsers/NativeTypeInfoParser.php:
```diff
@@ -0,0 +1,73 @@
+        $getter = sprintf(static::GETTER_FORMAT, ucfirst($property->getName()));
+        $class = $property->getDeclaringClass();
+        if ($class->hasMethod($getter)) {
+            return $method = $class->getMethod($getter);
```
unnecessary assignment (remove `$method =`)

dunglas commented on src/PropertyInfo/TypeInfoParsers/NativeTypeInfoParser.php:
```diff
@@ -0,0 +1,73 @@
+    public function getSetterParam(\ReflectionProperty $property)
+    {
+        $setter = $this->getSetter($property);
+        if ($setter === null or 1 !== $setter->getNumberOfRequiredParameters()) {
```
yoda style

dunglas commented on src/PropertyInfo/TypeInfoParsers/PHPTypeInfoParser.php:
```diff
@@ -0,0 +1,113 @@
+use PropertyInfo\Type;
+
+/**
+ * Property Extractor.
```
phpdoc

dunglas commented on src/PropertyInfo/TypeInfoParsers/PHPTypeInfoParser.php:
```diff
@@ -0,0 +1,113 @@
+ *
+ * @author Mihai Stancu <stancu.t.mihai@gmail.com>
+ */
+class PHPTypeInfoParser extends NativeTypeInfoParser
```
`PhpTypeInfoParser`

dunglas commented on src/PropertyInfo/Extractors/NativeExtractor.php:
```diff
@@ -0,0 +1,34 @@
+use PropertyInfo\TypeInfoParsers\PHPTypeInfoParser;
+
+/**
+ * Setter Extractor.
```
phpdoc

dunglas commented on src/PropertyInfo/Extractors/NativeExtractor.php:
```diff
@@ -0,0 +1,34 @@
+ */
+abstract class NativeExtractor implements TypeExtractorInterface
+{
+    /** @var  PHPTypeInfoParser|HHVMTypeInfoParser */
```
Should be `TypeInfoParserInterface`

dunglas replied:
And use the multiline style PHPDoc

dunglas commented:
Sorry I missed the first notification!
Awesome work, it was on my todo list for months.

I made some comments. Can you fix them and add some lines in the doc to explain how to enable PHP7 and HHVM support?

mihai-stancu commented:
Thanks!

I'll fix most issues today.

On the topic of `phpspec` vs `phpunit` I mentioned that I don't have prior experience with `phpspec` can we merge with the `phpunit` tests and I'll make another PR after I refactor them as `phpspec`?

dunglas commented:
I prefer to not multiply tools. Another solution is to port the few existing specs to PHPUnit.

mihai-stancu commented:
I commited the fixes but wasn't able to test the beforehand -- I have PHP7 compiled on another laptop -- hope they still run :-).

dunglas commented:
Travis has tested it for us!

dunglas commented:
Your previous comment seems to have been deleted. Strange.

dunglas commented on src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php:
```diff
@@ -0,0 +1,117 @@
+    {
+        $getter = $this->getGetter($property);
+
+        if (PHP_VERSION >= 7 and $getter !== null and $getter->hasReturnType()) {
```
`and` here too. The problem is precisely that many developers don't know the precedence trick and it can lead to bugs.
`and` and `or` are forbidden in the Symfony CS (and we follow Symfony CS) and are reported as violations by most quality checking tools including SensioLabs Insight and Scrutinizr (and we use both).

dunglas commented on src/PropertyInfo/Extractors/SetterExtractor.php:
```diff
@@ -10,48 +10,23 @@
 namespace PropertyInfo\Extractors;
```
Missing license header.

mihai-stancu replied:
I don't see the missing header in my files.

dunglas replied:
Sorry my bad it is here.

dunglas commented:
Tests fail with PHP7: https://travis-ci.org/dunglas/php-property-info/jobs/70364772

mihai-stancu commented:
Travis runs the `phpspec` not `phpunit`.

mihai-stancu commented:
Does the PhpDocExtractor read doc comments off of Properties as well as Getters and Setters?

It would stand to reason to refactor the PhpDocExtractor to use the TypeInfoParser logic as well.

dunglas commented:
would be cool

dunglas commented:
@mihai-stancu any news about that PR?

mihai-stancu commented:
I checked my `phpunit` tests yesterday, they worked correctly on all of my configs (php5, php7, hhvm). 

The `phpspec` tests were failing on an issue related to mock objects -- something about receiving an object which does not respect the expected interface -- I think it may be cause by returning null instead of empty array in some instances of my implementation.

I'll transcribe the `phpspec` tests into `phpunit` tomorrow and make sure those work.

mihai-stancu commented:
The `phpspec` tests from `master` fail with the same error when ran with PHP7. Which means the error wasn't induced by my PR branch.

I'll just go on with rewriting the tests as PHPUnit if that's alright with you.

dunglas commented:
Strange, it doesn't fail for others PR such as #2.
IMO you porting PHPUnit tests to PHPSpec will be easier (PHPSpec also use Prophecy).

dunglas commented on src/PropertyInfo/Tests/ExtractorTest.php:
```diff
@@ -0,0 +1,69 @@
+                    $this->assertEquals($expectedClass, $actualType->getClass());
+                }
+            } else {
+                $this->fail('Using ' . get_class($extractor) . ' type "' . $expectedType . '" resulted in an empty type list');
```
Use `sprintf` for such cases to improve readability.

dunglas commented:
For the data provider part: https://github.com/coduo/phpspec-data-provider-extension

dunglas commented:
Ok it only fails with PHP7. And it seems that PHPSpec isn't supporting PHP7 for now so it's OK.

mihai-stancu commented:
I've rewritten the phpspec tests as phpunit tests and added `phpunit` to `travis` & `composer` to be run automatically. PHP7 obviously still fails for `phpspec` but works fine on `phpunit`.

dunglas commented:
Can you fix new issues raised by Scrutinizr please? https://scrutinizer-ci.com/g/dunglas/php-property-info/inspections/c440da42-b348-429e-9ad1-04efa97435c8

mihai-stancu commented:
Hey,

I looked them over a bit, but didn't have the time to fix them. I'll get back to them tomorrow.

mihai-stancu commented:
I fixed the two big issues in Scrutinizer, the rest are either tests related or small.

mihai-stancu commented:
Bump.

dunglas commented:
Can you squash your commits?

mihai-stancu commented:
Squashed.

dunglas commented:
Thank you @mihai-stancu

dunglas commented:
@mihai-stancu your commit introduced a major Insight violations and some minors: https://insight.sensiolabs.com/projects/29b845cf-106d-45b4-99af-271f2dc3f7d5/analyses/23

Can you fix them please?

dunglas commented on composer.json:
```diff
@@ -23,11 +23,17 @@
       "PropertyInfo\\": "src/PropertyInfo"
     }
   },
+  "autoload-dev": {
```
Why is this necessary?

soullivaneuh replied:
It's for tests classes namespace definition AFAIK.

soullivaneuh replied:
Tests classes structure is not the same as source. It should be, isn't it?

dunglas replied:
Yes it must be.

mihai-stancu replied:
The purpose is to load the Tests namespace only for dev but it can be omitted.

dunglas replied:
Ok please remove it.

dunglas commented on phpunit.xml:
```diff
@@ -0,0 +1,5 @@
+<phpunit
```
Must be `phpunit.xml.dist`: https://github.com/dunglas/DunglasApiBundle/blob/master/phpunit.xml.dist

dunglas commented on src/PropertyInfo/TypeInfoParsers/HhvmTypeInfoParser.php:
```diff
@@ -0,0 +1,156 @@
+    {
+        $type = new Type();
+
+        if (isset(static::$types[$info])) {
```
the succession of `if`/`else` blocks can be avoided using `return` statements.

dunglas commented on src/PropertyInfo/TypeInfoParsers/HhvmTypeInfoParser.php:
```diff
@@ -0,0 +1,156 @@
+    {
+        $type = new Type();
+
+        if (false !== ($pos = strpos($info, '<'))) {
```
Same here, better use `return` than `if`/`else` to reduce the complexity.

dunglas commented on src/PropertyInfo/TypeInfoParsers/NativeTypeInfoParser.php:
```diff
@@ -0,0 +1,76 @@
+    public function getSetterParam(\ReflectionProperty $property)
+    {
+        $setter = $this->getSetter($property);
+        if (null === $setter or 1 !== $setter->getNumberOfRequiredParameters()) {
```
Please no `or`

soullivaneuh replied:
See #3 for this.

dunglas commented on src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php:
```diff
@@ -0,0 +1,100 @@
+    {
+        $getter = $this->getGetter($property);
+
+        if (PHP_VERSION >= 7 && $getter !== null && $getter->hasReturnType()) {
```
Please use yoda style (according to Symfony CS): `null === $getter`.

dunglas commented on src/PropertyInfo/TypeInfoParsers/HhvmTypeInfoParser.php:
```diff
@@ -0,0 +1,156 @@
+     */
+    public function getGetterReturnType(\ReflectionProperty $property)
+    {
+        $method = $this->getGetter($property);
```
Potential bug: `getGetter` can return `null`.

dunglas commented on src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php:
```diff
@@ -0,0 +1,100 @@
+     */
+    public function getGetterReturnType(\ReflectionProperty $property)
+    {
+        $getter = $this->getGetter($property);
```
Potential bug: `getGetter` can return `null`.

dunglas commented on src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php:
```diff
@@ -0,0 +1,100 @@
+    {
+        $param = $this->getSetterParam($property);
+        if (null !== $param) {
+            if (PHP_VERSION >= 7) {
```
Reduce the complexity of this method by removing `elseif` and `else` statements (`return` are sufficient)

dunglas commented on src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php:
```diff
@@ -0,0 +1,100 @@
+
+        if ($info === null) {
+            return null;
+        } elseif ($info === 'double') {
```
Same here, the `elseif` if not necessary because of the  `return` on the previous line.

dunglas commented on src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php:
```diff
@@ -0,0 +1,100 @@
+        $collection = false;
+        $class = null;
+
+        if ($info === null) {
```
Yoda style

dunglas commented on src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php:
```diff
@@ -0,0 +1,100 @@
+        $class = null;
+
+        if ($info === null) {
+            return null;
```
`return` (`null` is not necessary)

dunglas commented on src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php:
```diff
@@ -0,0 +1,100 @@
+            return null;
+        } elseif ($info === 'double') {
+            $info = 'float';
+        } elseif ($info === 'array') {
```
yoda style

dunglas commented on .travis.yml:
```diff
@@ -28,3 +27,4 @@ install:
 
 script:
   - vendor/bin/phpspec run
+  - vendor/bin/phpunit tests
```
`vendor/bin/phpunit` must be sufficient is the config is OK.

mihai-stancu replied:
If not configured in phpunit.xml then PHPUnit expects a `phpunit [options] <directory>`. I'll add the directory to the XML.

dunglas replied:
Yes, it's why you need a `phpunit.xml.dist` file (and the `.dist` allow custom configuration not versioned if necessary).

mihai-stancu replied:
Apparently  even if the testsuites are configured in the xml it still needs me to specify the directory (ex.: `.`).

dunglas commented on tests/DoctrineExtractors/DoctrineExtractorTest.php:
```diff
@@ -0,0 +1,109 @@
+{
+    public function extractorsDataProvider()
+    {
+        $properties = array(
```
Short array syntax.

soullivaneuh replied:
Will be auto-corrected with #4.

dunglas commented on tests/DoctrineExtractors/DoctrineExtractorTest.php:
```diff
@@ -0,0 +1,109 @@
+    public function extractorsDataProvider()
+    {
+        $properties = array(
+            array(
```
Short array syntax.

dunglas commented on tests/DoctrineExtractors/DoctrineExtractorTest.php:
```diff
@@ -0,0 +1,109 @@
+                'collection' => false,
+                'class'      => null,
+            ),
+            array(
```
same here

dunglas commented on tests/DoctrineExtractors/DoctrineExtractorTest.php:
```diff
@@ -0,0 +1,109 @@
+                'collection' => false,
+                'class'      => null,
+            ),
+            array(
```
same here

dunglas commented on tests/DoctrineExtractors/DoctrineExtractorTest.php:
```diff
@@ -0,0 +1,109 @@
+                'collection' => false,
+                'class'      => null,
+            ),
+            array('name' => 'binary', 'type' => 'resource', 'collection' => false, 'class' => null),
```
same here and below

dunglas commented on tests/NativeExtractors/Data/HackData.hh:
```diff
@@ -0,0 +1,340 @@
+    private int $int = 0;
+    private float $float = 0.0;
+    private string $string = '';
+    private array $array = array();
```
Short array syntax here and bellow.

dunglas commented:
Sorry @mihai-stancu my review was incomplete. Can you fix issues raised by Insight and my new comments please (you can open another PR or reopen this one).

mihai-stancu commented:
OK sure.

mihai-stancu commented:
Should we remove the `phpspec`s from the repo so as not to use two tools for the same purpose?

dunglas commented:
Yes `phpspec` can now be removed.

mihai-stancu commented:
@Soullivaneuh I'm getting a `PHP Fatal error:  Class 'Doctrine\Tests\Common\Cache\CacheTest'` I assume it's due recent changes. Any idea where it's coming from?

mihai-stancu commented:
Done. PR #5
